rstudio/rstudio
# Support ansi hyperlinks

### Intent

Adresses #1941 (replaces #10299)

### Approach

The hyperlink escape code described in https://iterm2.com/documentation-escape-codes.html are recognized, and the text between an opening anchor code (with a url) and a closing one (w/o) is displayed as a clickable link. 

The action when such a link is clicked depends on its 3 components `url`, `text`, `params` : 
 - if the url starts with `rstudio:viewer:` then the remaining is used as an url to open in the viewer: `rstudio:viewer:http://example.com`
 - if the url starts with `file://` the file is navigated to, `params` may contain `line=3:col=4` to navigate to line 3 and column 4 of the file. 
 - if the url is `rstudio:help` then `params$topic` and (optionally) `params$package` are used to open the relevant help page in the help pane. 
 - if the url is `rstudio:vignette`, same thing but for vignettes. 
 - if the url is `rstudio:run` then the text is sent to the console, params may contain parameters supported by `.rs.api.sendToConsole`
 - if the url starts with `rstudio:run:` then the remaining is sent to the console, e.g. `rstudio:run:head(mtcars)` 

For ease, the url may be generated by `cli::style_hyperlink()` (>= 3.1.1)

### Automated Tests

This has unit tests to assert that the right thing is written in the Dom: 

```java
   public void testHyperlinkNoParams()
   {
      PreElement ele = Document.get().createPreElement();
      VirtualConsole vc = getVC(ele);
      String testInput = "\u001b]8;;http://www.example.com\7text\u001b]8;;\7";
      vc.submit(testInput);
      String expected = "<a class=\"xtermUnderline xtermColor5 xtermHyperlink\" title=\"http://www.example.com\">text</a>";
      Assert.assertEquals(expected, ele.getInnerHTML());
   }

   public void testHyperlinkWithParams()
   {
      PreElement ele = Document.get().createPreElement();
      VirtualConsole vc = getVC(ele);
      String testInput = "\u001b]8;name1=value1:name2=value2;http://www.example.com\7text\u001b]8;;\7";
      vc.submit(testInput);
      String expected = "<a class=\"xtermUnderline xtermColor5 xtermHyperlink\" title=\"http://www.example.com\">text</a>";
      Assert.assertEquals(expected, ele.getInnerHTML());
   }
```

### QA Notes

```r
library(cli)
library(glue)

tf <- tempfile()
writeLines(c("bla bla", "yada yada yada"), tf)

links <- c(
  style_hyperlink('dplyr::summarise()', 'https://dplyr.tidyverse.org/reference/summarise.html' ),
  style_hyperlink('dplyr::summarise()', 'rstudio:viewer:https://dplyr.tidyverse.org/reference/summarise.html'),
  
  style_hyperlink('open file'                             , glue('file://{tf}')),
  style_hyperlink('open file at line 2'                   , glue('file://{tf}'), c(line = 2)),
  style_hyperlink('open file at line 2, column 6'         , glue('file://{tf}'), c(line = 2, col = 5)),
  
  style_hyperlink('?stats::rnorm' , "rstudio:help", c(topic = "rnorm", package = "stats")),
  style_hyperlink('?mean', "rstudio:help", c(topic = "mean")),
  
  style_hyperlink('vignette', "rstudio:vignette", c(topic = "type-size", package = "vctrs")),
  
  style_hyperlink("head(mtcars)", "rstudio:run", c(execute = FALSE)), 
  style_hyperlink("mtcars", "rstudio:run:head(mtcars)", c(animate = TRUE))
)

writeLines(links)
```

This should display links: 

![image](https://user-images.githubusercontent.com/2625526/150785038-4f004fde-b35d-457d-b8be-7002c0eb3517.png)

clicking on these links will: 

 - open the webpage in a browser
 - open the webpage in the viewer pane
 
 - navigate to the file
 - navigate to the 2nd line of the file
 - navigate to the 2nd line on column 6 of the file

 - open the help page for `stats::rnorm()`
 - open the help for `mean()`

 - open the type size vignette

 - put the code `head(mtcars)` in the console w/o executing it
 - put the code `head(mtcars)` in the console one character at a time, and then execute it. 


### Checklist
